### PR TITLE
Add couchdb clustering

### PR DESCRIPTION
**Pull request by style95**
This pr is to deploy couchdb cluster.
From the version 2.0, CouchDB officially supports clustering.

Notable changes are as follow:

### Image of CouchDB is changed to the one from Apache.

`klaemo/couchdb` is superseded by `apache/couchdb`.
Accordingly version is also updated to 2.1

### Coordinator node.

Since we need one coordinator node to setup CouchDB cluster, first host in db group becomes coordinator node.

### New environment variables for cluster.
Important factors for CouchDB cluster are the number of shards and the number of quorums.
Corresponding environment variables are introduced.

Note: Please be informed that, current version of `apache/couchdb:2.1` does not support these values, so I created a pull request for this.
Once this pr is merged, those environment variables will be available.
https://github.com/apache/couchdb-docker/pull/32

This pr also can address the problem reported by this issue: #2808

### CODE REVIEW

**Comment by style95** [comment]
Any more thing to do for this pr?

**Comment by rabbah** [comment]
@style95 as you shared on slack, the use of `db_num_read_quorum` and `db_num_write_quorum` I think is discouraged. Can you link the ansible related issue here and is there an update?

**Comment by style95** [comment]
@rabbah Let me explain the issues here.

In this pr, `apache/couchdb` is being used.
That image does not support environment variables for `n`, `q`, `r`, `w`, and that's the reason why I initially created this PR(apache/couchdb-docker#32).

And I got comment from couchdb community that,

1. `n` will be automatically configured while setting up clustering.
2.  `q` only matters when creating database and can be specified `?q=1` parameter.
3. Explicit `r` and `w` configurations are deprecated, instead couchdb will use half of the number of replicas for those two values.
(These can also be specified parameter) 
4. After setting up clustering, cluster configurations will be stored in `/opt/couchdb/etc/local.d`, so couchdb community highly recommend to use volume mapping of that directory.

In short, we don't need any those configurations.
Instead, we need to add `q` param when creating a database.

I will update the code accordingly.

One question is, would it be required to add volume mapping for configuration directory as well?
For this, we need to keep configuration directory and files in OW repo such as `etc/default.ini` `etc/default.d/*.ini` `etc/local.d` because there will be no default configuration when volume is mapped.

**Comment by style95** [comment]
@rabbah 
Code is updated accordingly.
Regarding `q` value, default is `8`.
I think it would be a good starting point, how do you think?

**Comment by style95** [comment]
It seems Travis error is not related with this pr.
Could anyone trigger Travis build again?

File: ansible/roles/couchdb/tasks/deploy.yml [hunk 1]
```diff
@@ -1,9 +1,9 @@
 ---
 # This role will run a CouchDB server on the db group
 
-- name: "Set node name to couchdb{{ groups['db'].index(inventory_hostname) }}"
+- name: "Set the coordinator to the first node"
   set_fact:
-    nodeName: "couchdb{{ groups['db'].index(inventory_hostname) }}"
+    coordinator: "{{ groups['db'][0] }}"
```
[the remark this answers is not in the document]
Comment (style95) [reply ansible/roles/couchdb/tasks/deploy.yml]: As you know, to setup cluster, we need only one coordinator node.
This is to get the first host among many hosts.
Is it possible to do the same thing with `ansible_host`?

For examaple, if there are two nodes running on same `ansible_host`, we need only one node.

File: ansible/roles/couchdb/tasks/deploy.yml [hunk 2]
```diff
@@ -21,25 +21,39 @@
...
-    image: klaemo/couchdb:{{ couchdb.version }}
+    image: apache/couchdb:{{ couchdb.version }}
     state: started
     recreate: true
     restart_policy: "{{ docker.restart.policy }}"
     volumes: "{{volume_dir | default([])}}"
     ports:
       - "{{ db_port }}:5984"
+      - "4369:4369"
+      - "9100:9100"
```
[the remark this answers is not in the document]
Comment (style95) [reply ansible/roles/couchdb/tasks/deploy.yml]: I want to discuss this further.
Please refer to this: http://docs.couchdb.org/en/latest/cluster/setup.html#firewall

Those two ports are used when Erlang VM finds other nodes.
APAIK, 4369 is reserved to find other nodes, and 9100 is reserved to use for communication among nodes.

Normally, erlang VM uses random ports between 9100~9200, but it is forced to use 9100 only in `apache/couchdb`.

Regarding 9100, I think we can change it, but we need to update `apache/couchdb` image.
That should be different on every nodes, I am not sure it works in that situation.
You can refer to this: https://github.com/apache/couchdb-docker/blob/master/2.1.0/vm.args#L14

And with regard to 4369, since it is reserved at Erlang VM layer, I am not sure we can change it.

@cbickel Do you have any idea?

**Comment by style95** [comment]
@cbickel any comments on this?

File: ansible/roles/couchdb/tasks/deploy.yml [hunk 3]
```diff
@@ -49,9 +63,48 @@
   retries: 12
   delay: 5
 
+- name: enable the cluster setup mode
+  uri:
+    url: "{{ db_protocol }}://{{ ansible_host }}:{{ db_port }}/_cluster_setup"
+    method: POST
+    body: >
+        {"action": "enable_cluster", "bind_address":"0.0.0.0", "username": "{{ db_username }}", "password":"{{ db_password }}", "port": {{ db_port }}, "node_count": "{{ num_instances }}", "remote_node": "{{ ansible_host }}", "remote_current_user": "{{ db_username }}", "remote_current_password": "{{ db_password }}"}
```
Comment (cbickel) [inline ansible/roles/couchdb/tasks/deploy.yml]: What about using
```
{{ groups['db'] | length }}
```
Instead of defining `num_instances` on the top?

**Comment by cbickel** [comment]
This is a great Pull request.
The only thing, that should be there (if it is technically possible) is, that we should have the ability to deploy two instances on the same machine. With this feature, testing would be much easier. We have it for the controllers and invokers as well.

And I have an additional question. Is the intention of this PR only, to share data across multiple instances? Wouldn't it be great, if the requests against the database would be distributed across the instances? Or do you have in mind to do this later?

**Comment by style95** [comment]
Agree. It would be great if we can deploy more than one instances on a single machine for testability.

Regarding the data sharing across multiple instances, do we need any further configurations to distribute data?

AFAIK, from CouchDB 2.0, the number of nodes who are in charge of specific shards will be decided based on `n`(normally recommended value is 3).
For example, if we have 5 nodes, and `n` is 3, and let's say node 1,2,3 are in charge of some data, and if client sent request to node 4, it will automatically redirect the request to one of node 1,2,3 and they will handle the requests properly.

Based on the value of `w`, `r`, it will respond to the clients.
So no matter who receives the request, client will get proper response.
Did I understand your question correctly?
If I misunderstood about couchdb, please rectify me.

If you are mentioning about load-balancing(using single endpoint address) among multi nodes for HA, currently i have no idea. We may locate load-balancer in front of them or, other components to use list of endpoints.

**Comment by cbickel** [comment]
Ok, I didn't know yet, that if you send a request to instance 1, that it could be handled by instance2. 

I think for HA we need one of your proposed ideas. I just wanted to ask, if you want to address that problem already in this PR, or maybe later.
But for this PR I'm totally fine without loadbalancing to the different nodes.

**Comment by style95** [comment]
Yes I think it would be better to address load-balancing problem in separate pr.

Regarding `epmd` port, `4369`, initially I thought we can change the port using `ERL_EPMD_PORT`: http://erlang.org/doc/man/epmd.html.

But for example, if we configure the port as `4369` on first node and `4370` for second node, first node will try to find other node using `4369` port but on second node, only `4370` is opened, so look-up will fail.

I am looking for the configuration to segregate its binding port(`4369`) and communication port(`4370`). But I am not sure it is feasible.

**Comment by style95** [comment]
It seems, all Erlang nodes must use same epmd port.
https://stackoverflow.com/questions/40189097/erlang-epmd-connects-to-other-host-with-non-default-epmd-port

It means, we cannot run a couchdb cluster on a single machine.

As of now I am trying the following scenario.

1. use same epmd port(`4369`) for all couchdb nodes.
2. do not map that port to host (remove `-p 4369:4369`).
3. trying to form a cluster.

Since both nodes will be running on a single machine, they will be in the same subnet.
They might be able to communicate with each other using their container IPs.

So though port is not opened in host layer, they may be able to connect to other node using `172.17.0.2:4369`.
For this, when we deploy couchdb, we need to specify its container IPs.
(currently, we are specifying host IP).

Since it assumes nodes are running on same subnet, it won't work on distributed environments.
So we need to differentiate local deployment with distributed deployment.


Is it worth to take this approach?

**Comment by style95** [comment]
Let me share the discussion with @cbickel.
It would be better for us to have test cases for HA or failover of CouchDB.
But it's not that simple and seems to require many dirty changes to run multiple couchdb on a single machine.

So it could be an option that running a test case in case there are more than 1 couchdb deployed(in distributed environment).
Same approach is already taken for controllers.

I will look into `ShootComponentsTests.scala` and apply similar way for couchdb as well.

**Comment by cbickel** [comment]
One idea for one of the tests could be:
- Write documents to the first DB
- Shoot the first DB
- Check that the documents are available (maybe also some view results) on the second db
- Write some documents to second DB
- Start first DB again and wait until the first DB is up again
- Check that the new documents are replicated back to the first database

Do you have some more ideas?

**Comment by chetanmeh** [comment]
@style95 What would be consistency support in clustered setup. Would it be possible to read-your-own-writes?

Reading [here][1] indicates that for reads and writes it would use quorum which should allow one to read its own write

> The number of copies of a document with the same revision that have to be read before CouchDB returns with a 200 is equal to a **half of total copies of the document plus one**. It is the same for the number of nodes that need to save a document before a write is returned with 201

However its not clear what is the semantics for queries or view access. Per [here][2] it looks like queries would not be using quorum

> The coordinator waits for a given number of responses to arrive from the nodes it contacted. The number of responses it waits for differ based on the type of request. Reads and writes wait for two responses. **A view read only waits for one.**

So in clustered setup it can happen that an action added in one step is not visible in query made in subsequent request?

[1]: http://docs.couchdb.org/en/latest/cluster/theory.html
[2]: https://dx13.co.uk/articles/2015/10/19/couchdb-20s-read-and-write-behaviour-in-a-cluster.html

**Comment by rabbah** [comment]
Correct: views are subject to eventual consistency and may exclude a just added/modified action (for example) until all nodes have replicated and indexed.

**Comment by chetanmeh** [comment]
Thanks @rabbah  for confirming the eventual consistent nature for queries. 

Next question - Does Couchdb supports specifying the sharding key? Or it always shards on the basis of _id key only. In absence of sharding key all queries would need to query all shards which may have some impact on performance

**Comment by style95** [comment]
@chetanmeh 
AFAIK, there is no support to specify the custom sharding key.
What kinds of performance increment do you expect with custom sharding key?

If we do view query, we should access all relevant shards to create a combined result.
That's the reason why, it is generally recommended to have small `q` value for read-intensive DB and large value for write-intensive DB.

If I misunderstood, kindly correct me.

**Comment by chetanmeh** [comment]
> What kinds of performance increment do you expect with custom sharding key?

I am exploring support for running OW on Mongo and Cosomosdb which support specifying custom partition key. Looking at views used by Couchdb I see following types of queries
 
* Activations
  * byDateWithLogs - [doc.start]
  * cleanUpActivations - [doc.start]
  * activations - [doc.namespace, doc.start]
* Whisks
  * [root namespace, date] - For each of rules, packages, actions, triggers

From my understanding of flow the "whisks" are mostly read. In a sharded setup this may benefit by using "root namespace" as sharding (partition key). In such a case the query would only need to be evaluated on single shard and avoid hitting all the shards for the whisks access. 

This would be specially useful for databases where sharded queries are done from client which may add higher latency. Compared to this Couchdb proxies the queries and talks to all the shards on behalf of client. As such calls happen within same data center they may incur less latency. 

So I am interested in this PR to see how to model OW entities on other sharded storage systems

**Comment by style95** [comment]
So you are thinking of assigning a shard per namespace, right?
Great idea. It would be great for small scale and latency.

But if there are huge number of data, one shard may not be enough.
Since shard is a unit of distribution and replication, too big shard is not good for performance.
As data grows, at some scale we need to rebalance shards but if we use namespace as a partition key, how can we rebalance the shards into small shards?

And accessing to shards are proceeded in parallel.
Generally recommended value for `n` is 3 as so, `w` and `r` will be 2.
That means the number of shards to access is 2.
Though we decrease the access time to 1, there might not be huge latency improvement.

And in my experience, storing activation has more huge impact on performance than reading whisk entity because there is cache layer in controller.

**Comment by chetanmeh** [comment]
> But if there are huge number of data, one shard may not be enough.
Since shard is a unit of distribution and replication, too big shard is not good for performance.
As data grows, at some scale we need to rebalance shards but if we use namespace as a partition key, how can we rebalance the shards into small shards?

Ack. But it depends on kind of data being stored. This scheme may be ok for whisks where one namespace may contain say 1M entities (which can be stored in single shard). This may not work fine for activations where number of activations may of the order of 10-100M. So use of sharding key would depend on type of objects being stored

> And accessing to shards are proceeded in parallel.
Generally recommended value for n is 3 as so, w and r will be 2.
That means the number of shards to access is 2.
Though we decrease the access time to 1, there might not be huge latency improvement.

This is fine for read and write of single instance. However for queries [all shards would need to be queried][1]

> The Query request handler must ask every node that hosts a replica of a shard in the database for their answer to the query. Depending on how the shard replicas are distributed across the cluster, this might be all nodes in the cluster, or only a subset.

> And in my experience, storing activation has more huge impact on performance than reading whisk entity because there is cache layer in controller.

Is the cache layer used for queries also or its only used for individual entity access i.e. lookups by docId?

[1]: https://dx13.co.uk/articles/2016/1/31/understanding-couchdb-20-and-cloudants-indexing-behaviour-in-a-cluster.html

**Comment by style95** [comment]
@chetanmeh 
Yes. There might not be such huge number of data in one namespace.
But what if data grows by any possibility, how can we handle that?
And if there are not much data, do you expect it will cause any trouble in terms of performance with such small size of data?

Second, query is only used to list entities.
All other APIs, such as get/update/delete and even invoke an action uses normal document get/put api.(surely they can take advantage of cache.)

And I can't expect such huge number of `list` API calls.
As you mentioned, your approach can improve latency of `list` API.
But normally `list` API is called by user rather than any systematic call because list is to see what kinds of entities are there. So there may not be 10K or 100K TPS calls.

That's the reason why I think action invocation(which includes storing activations) will affect more on the performance and incline to scalable approach.

**Comment by chetanmeh** [comment]
> But normally list API is called by user rather than any systematic call because list is to see what kinds of entities are there. So there may not be 10K or 100K TPS calls.
That's the reason why I think action invocation(which includes storing activations) will affect more on the performance and incline to scalable approach.

Makes sense. Would keep that in mind when modeling for other dbs

> And, I could not find any support for custom partition key.

Thanks for confirming. So for couchdb the partition key decision is moot!

**Comment by rabbah** [comment]
The queries (list operations) are not cached today. They could be since we allow stale results and queue indexing operations in couch as configured. In general as already noted the bulk of the load isn’t on crud or query operations.

**Comment by style95** [comment]
I have tested this pr is working with a couchdb cluster and a single couchdb setup.

**Comment by style95** [comment]
Travis complains there is no `db_local.ini`, is it because of this PR?

File: tests/src/test/scala/ha/ShootComponentsTests.scala [hunk 4]
```diff
@@ -63,18 +60,61 @@ class ShootComponentsTests extends FlatSpec with Matchers with WskTestHelpers wi
...
+    val response = Try {
+      Http().singleRequest(HttpRequest(uri = s"http://$host:$port/ping")).futureValue
+    }.toOption
+
+    response.map { res =>
+      (res.status, Unmarshal(res).to[String].futureValue)
+    }
+  }
+
+  def pingDB(host: String, port: Int) = {
```
Comment (cbickel) [inline tests/src/test/scala/ha/ShootComponentsTests.scala]: what do you think about using one ping method and passing the `path` as an argument.
